doc/lsst.daf.butler/formatters.rst

```diff
@@ -0,0 +1,316 @@
+
+Assemblers
+==========
+
+.. note::
+
+  The base class is called CompositeAssembler because it was first developed for composites.
+  Now might be a good time to rebrand it since non-composites need one if they use read parameters.
+
+Every `~lsst.daf.butler.StorageClass` that defines read parameters or components (read/write or read) must also specify an `~lsst.daf.butler.CompositeAssembler` class.
+This class should inherit from the `~lsst.daf.butler.CompositeAssembler` base class.
+
+Composite Disassembly
+^^^^^^^^^^^^^^^^^^^^^
+
+A composite is declared by specifying components in the `~lsst.daf.butler.StorageClass` definition.
+Assemblers must provide at minimum a `~lsst.daf.butler.CompositeAssembler.getComponent()` method to enable a specific component to be extracted from the composite Python type.
+Datastores can be configured to prefer to write composite datasets out as the individual components and to reconstruct the composite on read.
+This can lead to more efficient use of datastore bandwidth (especially an issue for an S3-like storage rather than a local file system) if a pipeline always takes as input a component and does not require the full dataset or if a user in the science platform wants to retrieve the metadata for many datasets.
+To allow this the assembler subclass must provide `~lsst.daf.butler.CompositeAssembler.assemble()` and `~lsst.daf.butler.CompositeAssembler.disassemble()`.
+
+Datastores can be configured to always disassemble composites or never disassemble them.
+Additionally datastores can choose to only disassemble specific storage classes or dataset types.
+
+.. warning::
+
+  Composite disassembly implicitly assumes that an identical Python object can be created from the disassembled components.
+  If this is not true, the components should be declared read-only and disassembly will never be attempted.
+
+Read-only Components
+^^^^^^^^^^^^^^^^^^^^
+
+Just as for components of a composite, if a storage class defines read-only components, it must also specify an assembler to support the calculation of that derived component.
+This should be implemented in the `~lsst.daf.butler.CompositeAssembler.getComponent()` method.
+
+Additionally, if the storage class refers to a composite, the datastore can be configured to disassemble the dataset into discrete artifacts.
+Since read-only components are derived and are not persisted themselves the datastore needs to be told which component should be used to calculate this derived quantity.
+To enable this the assembler must implement `~lsst.daf.butler.CompositeAssembler.selectResponsibleComponent()`.
+This method is given the name of the read-only component and a list of all available persisted components and must return one and only one relevant component.
```

Comment:
I put it in the assembler so that the assembler can make a dynamic choice based on the available components. I also don't know how common composite disassembly is going to become so it wasn't clear to me that I really needed more complexity in the storage class configuration. The assembler is the thing that has to calculate the "computed component" so it seemed reasonable to expect the assembler to know which component it needs to use for that.